### ISSUE TITLE
Much Needed: open-source realtime database for "mobile apps"

### ISSUE THREAD

**Issue by Rajan** [rethinkdb/rethinkdb#5239]
There are many real-time solutions for "realtime web" (arguably not comparable to RethinkDB). Ideally, It'd be great to have something that works seamelessly on "mobile apps" ideally with offline-first and real-time sync abilities. When can we expect great iOS/Android libraries? .. which can make rethinkdb a true much-needed Firebase alternative. Hoping to contribute to make this happen. Thanks in advance.

**Comment by deontologician** [comment]
ping @coffeemug

**Comment by coffeemug** [comment]
@Rajan -- we're working on it (see https://github.com/rethinkdb/rethinkdb/issues/3711#issuecomment-151639571).

I've added you to the project, would love your feedback!

**Comment by Rajan** [comment]
Thanks @coffeemug 
There's a project along the lines of realtime-web-apps: http://deepstream.io/
Also, http://telepat.io/ is doing a fair job using couchbase and other things but too many tools.
Ideally, I'd like to see the Fusion developed as "mobile first" framework rather than web-first which seems to be the case. Being great mobile first platform should have these:

* Offline-first and real-time sync Swift/Java sdks for iOS/Android
* Out of the box watching of "changefeeds" from those libraries.
* Passportjs supported user creation and authentication

Some iOS/Swift work is done here, may be we can reuse/rewrite:
https://github.com/dorjeozer/LoopBack-Swift
https://github.com/pixelspark/rethink-swift
Later, we can have the same functionality with Android sdk.

**Comment by Rajan** [comment]
There are a large number of realtime-web frameworks these days. They all work with socket.io and browser and be able to do similar things is less likely gain great attention/traction.

I do like the RethinkDB approach to building "realtime" fundamentally into the database, it is hard for front-end or web developers to relate to it and choose a new framework based on the same story. It is particularly hard because it is not the code that they're going "direclty" work with and hence they won't value it as much.

An alternative strategy would be plug into every popular open-source framework and turn it into "Realtime" mobile apps framework. e.g. We can get started with LoopBack.io and create a serious Loopback-RethinkDB-Realtime and add "changefeed" events/methods to their existing iOS/Android SDKs by creating RethinkDB forks of their sdks. It'd be relatively easy to attract existing node.js/loopback developers to explore RethinkDB as a truly "Realtime" alternative to MongoDB. It can work because Mongo doesn't support joins and they've to do very little work to use RethinkDB instead of Mongo.

Later, we can contribute ios/android sdks to deepstream (they're already building on RethinkDb).

**Comment by danielmewes** [comment]
@Rajan I've opened an issue for offline support in the other repository, and will open a second one for better Android/iOS integration.

The Fusion protocol is designed to allow for fairly straight-forward integration into a wide number of frameworks. I think this is something we will look into fairly soon after we've got the JavaScript API fully working.

I'm going to close this issue in favor of the other ones.
(the other repository isn't public yet, but will be fairly soon once we have something to release)
